tree.go

```diff
@@ -449,75 +451,39 @@ func (tree *Tree) newnode() (p *node) {
 	return &(tree.alloc[ln])
 }
 
-func loadip4(ipstr []byte) (uint32, error) {
-	var (
-		ip  uint32
-		oct uint32
-		b   byte
-		num byte
-	)
-
-	for _, b = range ipstr {
-		switch {
-		case b == '.':
-			num++
-			if 0xffffffff-ip < oct {
-				return 0, ErrBadIP
-			}
-			ip = ip<<8 + oct
-			oct = 0
-		case b >= '0' && b <= '9':
-			oct = oct*10 + uint32(b-'0')
-			if oct > 255 {
-				return 0, ErrBadIP
-			}
-		default:
-			return 0, ErrBadIP
-		}
-	}
-	if num != 3 {
-		return 0, ErrBadIP
-	}
-	if 0xffffffff-ip < oct {
-		return 0, ErrBadIP
-	}
-	return ip<<8 + oct, nil
-}
+func parseCIDR(cidr []byte) (net.IP, net.IPMask, error) {
+	var address netip.Addr
+	var prefixLength int
 
-func parsecidr4(cidr []byte) (uint32, uint32, error) {
-	var mask uint32
-	p := bytes.IndexByte(cidr, '/')
-	if p > 0 {
-		for _, c := range cidr[p+1:] {
-			if c < '0' || c > '9' {
-				return 0, 0, ErrBadIP
-			}
-			mask = mask*10 + uint32(c-'0')
+	// Check for '/' to determine if this is a single IP or prefix. This is the same approach used by net.ParseCIDR
+	if bytes.IndexByte(cidr, '/') < 0 {
+		addr, err := netip.ParseAddr(string(cidr))
+
+		if err != nil {
+			return nil, nil, err
 		}
-		mask = 0xffffffff << (32 - mask)
-		cidr = cidr[:p]
+
+		address = addr
+		prefixLength = address.BitLen()
 	} else {
-		mask = 0xffffffff
-	}
-	ip, err := loadip4(cidr)
-	if err != nil {
-		return 0, 0, err
-	}
-	return ip, mask, nil
-}
+		prefix, err := netip.ParsePrefix(string(cidr))
 
-func parsecidr6(cidr []byte) (net.IP, net.IPMask, error) {
-	p := bytes.IndexByte(cidr, '/')
-	if p > 0 {
-		_, ipm, err := net.ParseCIDR(string(cidr))
 		if err != nil {
 			return nil, nil, err
 		}
-		return ipm.IP, ipm.Mask, nil
+
+		address = prefix.Addr()
+		prefixLength = prefix.Bits()
 	}
-	ip := net.ParseIP(string(cidr))
-	if ip == nil {
-		return nil, nil, ErrBadIP
+
+	if address.Is4In6() {
+		address = address.Unmap()
+		prefixLength -= 8 * (net.IPv6len - net.IPv4len)
```

Comment:
for future clarity: I know what this is doing but I'd drop a tiny comment to make clear it's adjusting by 96 bits (128 -> 32)